pkg/gatewayserver/io/io.go

```diff
@@ -734,3 +735,49 @@ func (c *Connection) discardRepeatedUplink(up *ttnpb.UplinkMessage) bool {
 	}
 	return shouldDiscard
 }
+
+// DropDuplicatesByRSSI returns the given list of gateway uplink messages after discarding
+// duplicates by RSSI. Two gateway uplink messages are considered duplicates if the RawPayload
+// is identical, and the RSSI values differ. In these cases, only the gateway uplink message
+// with the highest RSSI will be included in the result.
+//
+// DropDuplicatesByRSSI will allocate a new list of uplink messages, but will not copy the uplink
+// messages themselves.
+func DropDuplicatesByRSSI(uplinks []*ttnpb.UplinkMessage) []*ttnpb.UplinkMessage {
+	// Quick out
+	if len(uplinks) < 2 {
+		return uplinks
+	}
+
+	// Create map for max RRSI by payload (base64)
+	maxRSSI := make(map[string]float32, len(uplinks))
+	for _, uplink := range uplinks {
+		md := uplink.GetRxMetadata()
+		if len(md) == 0 {
+			continue
+		}
+		key := base64.StdEncoding.EncodeToString(uplink.GetRawPayload())
+		if rssi, ok := maxRSSI[key]; !ok || rssi < md[0].RSSI {
+			maxRSSI[key] = md[0].RSSI
+		}
+	}
+	if len(maxRSSI) == len(uplinks) {
+		return uplinks
+	}
+
+	// Discard uplinks when the same payload with a larger RSSI was found
+	deduplicated := make([]*ttnpb.UplinkMessage, 0, len(maxRSSI))
```

Comment:
Capacity needs to be `len(uplinks)`; there can be more uplinks with no metadata than deduplicated uplinks with RSSI